src/infra/core/src/services/metadata_query_service_impl.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Utc};
+use dill::*;
+use internal_error::{InternalError, ResultIntoInternal};
+use kamu_core::*;
+use opendatafabric::{self as odf, AsTypedBlock};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[interface(dyn MetadataQueryService)]
+pub struct MetadataQueryServiceImpl {}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl MetadataQueryService for MetadataQueryServiceImpl {
```

Comment:
Would these methods be better suited in `MetadataChainExt` extension trait?

To me a service encapsulates complex interactions between several components, but here we have nothing but a stateless wrapper on top of `MetadataChain`'s visitor interface.

`MetadataChain` is a place where we also validate complex domain rules about what sequences of events are valid an which aren't (e.g. two active push sources with same name) so I don't think an argument that this is some "higher-level logic" works in this case either.

Reply:
I'm not sure. I have a feeling that for high level queries to dataset's current state, we shouldn't be considering that there is some metadata chain behind it. We just want to know key characteristics of the dataset's current state. The fact that this maps to `MetadataChain` algorithm is an implementation detail. Besides, these questions could be answered with database in future.

Reply:
`MetadataChainExt` deals with a much lower level of abstraction. This is mostly about visiting the blocks one way or another,  emphasizing explicit knowledge of chain structure.